File: backend/models/user.model.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;

const UserSchema = new Schema({
    name: { 
        type: String, 
        required: true 
    },
    email: { 
        type: String, 
        required: true, 
        unique: true,
        trim: true,
        lowercase: true
    },
    password: {
        type: String,
        required: true
    },
    phone: {
        type: String,
        // Bỏ required để user Google có thể đăng ký mà không cần SĐT ban đầu
    },
    address: {
        type: String,
        // Bỏ required để user Google có thể đăng ký mà không cần địa chỉ ban đầu
    },
    location: {
        type: {
          type: String,
          enum: ['Point'],
          default: 'Point'
        },
        coordinates: {
          type: [Number], // [longitude, latitude]
          default: [0, 0]
        }
    },
    role: {
        type: String,
        enum: ['customer', 'delivery_personnel', 'admin'],
        default: 'customer'
    },
    // Các trường dành riêng cho người giao hàng
    vehicle: {
        type: Schema.Types.ObjectId,
        ref: 'Vehicle'
    },
    availabilityStatus: {
        type: String,
        enum: ['available', 'unavailable', 'on_delivery'],
        default: 'unavailable'
    },
    avatarUrl: { // Thêm trường avatar
        type: String,
        default: ''
    }
// Dấu ngoặc đóng của schema phải ở đây, chỉ có một khối timestamps
}, {
    timestamps: true
});

// UserSchema.index phải được gọi sau khi đã định nghĩa xong Schema
UserSchema.index({ location: '2dsphere' });

module.exports = mongoose.model('User', UserSchema);

File: frontend_app/lib/shared/models/user_model.dart
// File: lib/shared/models/user_model.dart
class UserModel {
  final String id;
  final String name;
  final String email;
  final String? phone;      // <<-- SỬA: Cho phép null
  final String? address;    // <<-- SỬA: Cho phép null
  final String? avatarUrl;
  final String role;

  UserModel({
    required this.id,
    required this.name,
    required this.email,
    required this.role,
    this.phone,         // <<-- SỬA: Không còn là required
    this.address,       // <<-- SỬA: Không còn là required
    this.avatarUrl,
  });

  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      // Cung cấp giá trị mặc định an toàn cho các trường không thể null
      id: json['_id'] ?? '', // Nếu _id là null, trả về chuỗi rỗng
      name: json['name'] ?? 'Người dùng mới', // Nếu name là null, trả về giá trị mặc định
      email: json['email'] ?? '', // Nếu email là null, trả về chuỗi rỗng
      role: json['role'] ?? 'customer',

      // Các trường nullable có thể nhận giá trị null trực tiếp
      phone: json['phone'],
      address: json['address'],
      avatarUrl: json['avatarUrl'],
    );
  }
}